components/ads-client/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+use std::collections::{HashMap, HashSet};
+
+use error::ApiResult;
+use error::{
+    BuildPlacementsError, BuildRequestError, ComponentError, RecordClickError,
+    RecordImpressionError, ReportAdError, RequestAdsError,
+};
+use error_support::handle_error;
+use instrument::TrackError;
+use mars::{DefaultMARSClient, MARSClient};
+use models::{AdContentCategory, AdRequest, AdResponse, IABContentTaxonomy, MozAd};
+use parking_lot::Mutex;
+use uuid::Uuid;
+
+mod error;
+mod instrument;
+mod mars;
+mod models;
+mod test_utils;
+
+uniffi::setup_scaffolding!("adsclient");
+
+/// Top-level API for the mac component
+#[derive(uniffi::Object)]
+pub struct MozAdsClient {
+    inner: Mutex<MozAdsClientInner>,
+}
+
+impl Default for MozAdsClient {
+    fn default() -> Self {
+        Self {
+            inner: Mutex::new(MozAdsClientInner::new()),
+        }
+    }
+}
+
+#[uniffi::export]
+impl MozAdsClient {
+    #[uniffi::constructor]
+    pub fn new() -> Self {
+        Self {
+            inner: Mutex::new(MozAdsClientInner::new()),
+        }
+    }
+
+    #[handle_error(ComponentError)]
+    pub fn request_ads(
+        &self,
+        moz_ad_configs: Vec<MozAdsPlacementConfig>,
+    ) -> ApiResult<HashMap<String, MozAdsPlacement>> {
+        let inner = self.inner.lock();
+        let placements = inner
+            .request_ads(&moz_ad_configs)
```

Comment:
I'm not sure how your component is actually going to be used, but the locking here seems like it might not be ideal given the mutex is held while doing network IO. For example, on a very slow network, I could imagine subsequent calls to `record_impression` could end up starving the app of worker threads.

(There's actually no great solution here - not holding the lock while doing the network request just moves the bottle-neck somewhere else.I suspect a queue of requests might be where you end up needing to be?) Or ultimately with an async API?